# Overfitting a Neural Net
In this notebook I will use the Titanic dataset from Kaggle to show an example of overfitting an artificial neural network. The model will then be significantly reduced but will be a much more robust net.

In [ ]:
import numpy as np
import pandas as pd
import tensorflow as tf
from sklearn.impute import SimpleImputer
import warnings
warnings.filterwarnings('ignore')

In [ ]:
from google.colab import drive
drive.mount('/content/drive/', force_remount=True)

Go to this URL in a browser: https://accounts.google.com/o/oauth2/auth?client_id=947318989803-6bn6qk8qdgf4n4g3pfee6491hc0brc4i.apps.googleusercontent.com&redirect_uri=urn%3aietf%3awg%3aoauth%3a2.0%3aoob&response_type=code&scope=email%20https%3a%2f%2fwww.googleapis.com%2fauth%2fdocs.test%20https%3a%2f%2fwww.googleapis.com%2fauth%2fdrive%20https%3a%2f%2fwww.googleapis.com%2fauth%2fdrive.photos.readonly%20https%3a%2f%2fwww.googleapis.com%2fauth%2fpeopleapi.readonly

Enter your authorization code:
··········
Mounted at /content/drive/


## Importing the dataset

In [ ]:
data = pd.read_csv('/content/drive/My Drive/Colab Notebooks/train.csv')
dummies = pd.get_dummies(data.Sex)
data = pd.concat([data,dummies], axis = 'columns')
dummies2 = pd.get_dummies(data.Embarked)
data = pd.concat([data,dummies2], axis = 'columns')
data.drop(['PassengerId','Name','Ticket','Cabin','Sex','male','Embarked','S'], axis = 1, inplace = True)
x = data.iloc[:,1:9].values
y = data.iloc[:,0].values

In [ ]:
data2 = pd.read_csv('/content/drive/My Drive/Colab Notebooks/test.csv')
dummies = pd.get_dummies(data2.Sex)
data2 = pd.concat([data2,dummies], axis = 'columns')
dummies2 = pd.get_dummies(data2.Embarked)
data2 = pd.concat([data2,dummies2], axis = 'columns')
data2.drop(['PassengerId','Name','Ticket','Cabin','Sex','male','Embarked','S'], axis = 1, inplace = True)
x_test = data2.iloc[:,:].values

## Missing Data

In [ ]:
impute = SimpleImputer(missing_values = np.nan, strategy = 'mean')
impute.fit(x[:,1:3])
x[:, 1:3] = impute.transform(x[:, 1:3])
impute.fit(x_test[:,0:3])
x_test[:,0:3] = impute.transform(x_test[:, 0:3])
impute.fit(x_test[:,3:5])
x_test[:,3:5] = impute.transform(x_test[:, 3:5])

## Feature Scaling

In [ ]:
from sklearn.preprocessing import StandardScaler
sc = StandardScaler()
x = sc.fit_transform(x)
x_test = sc.transform(x_test)

## Building and training the ANN

In [6]:
ann = tf.keras.models.Sequential()

ann.add(tf.keras.layers.Dense(units=100, activation='relu', input_dim = 8))

ann.add(tf.keras.layers.Dense(units=100, activation='relu'))
ann.add(tf.keras.layers.Dense(units=100, activation='relu'))
ann.add(tf.keras.layers.Dense(units=100, activation='relu'))
ann.add(tf.keras.layers.Dense(units=100, activation='relu'))
ann.add(tf.keras.layers.Dense(units=100, activation='relu'))
ann.add(tf.keras.layers.Dense(units=100, activation='relu'))
ann.add(tf.keras.layers.Dense(units=100, activation='relu'))
ann.add(tf.keras.layers.Dense(units=100, activation='relu'))
ann.add(tf.keras.layers.Dense(units=100, activation='relu'))

ann.add(tf.keras.layers.Dense(units=1, activation='sigmoid'))

ann.compile(optimizer = 'adam', loss = 'binary_crossentropy', metrics = ['accuracy'])

ann.fit(x, y, batch_size = 32, epochs = 500)

Epoch 1/500
28/28 [==============================] - 0s 4ms/step - loss: 0.5869 - accuracy: 0.7172
Epoch 2/500
28/28 [==============================] - 0s 3ms/step - loss: 0.4382 - accuracy: 0.8227
Epoch 3/500
28/28 [==============================] - 0s 4ms/step - loss: 0.4038 - accuracy: 0.8350
Epoch 4/500
28/28 [==============================] - 0s 3ms/step - loss: 0.4057 - accuracy: 0.8361
Epoch 5/500
28/28 [==============================] - 0s 4ms/step - loss: 0.3912 - accuracy: 0.8283
Epoch 6/500
28/28 [==============================] - 0s 3ms/step - loss: 0.3900 - accuracy: 0.8451
Epoch 7/500
28/28 [==============================] - 0s 3ms/step - loss: 0.3876 - accuracy: 0.8429
Epoch 8/500
28/28 [==============================] - 0s 4ms/step - loss: 0.3809 - accuracy: 0.8406
Epoch 9/500
28/28 [==============================] - 0s 3ms/step - loss: 0.3780 - accuracy: 0.8406
Epoch 10/500
28/28 [==============================] - 0s 4ms/step - loss: 0.3681 - accuracy: 0.8474
Epoch 11/

The accuracy of this model looks really good, over 92%, but we will see that this is actually an overfit net.

## Splitting into training and test sets to understand the overfitting problem

In [ ]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(x, y, test_size = 0.2, random_state = 0)

Traing the model on the split data:

In [8]:
ann = tf.keras.models.Sequential()

ann.add(tf.keras.layers.Dense(units=100, activation='relu', input_dim = 8))

ann.add(tf.keras.layers.Dense(units=100, activation='relu'))
ann.add(tf.keras.layers.Dense(units=100, activation='relu'))
ann.add(tf.keras.layers.Dense(units=100, activation='relu'))
ann.add(tf.keras.layers.Dense(units=100, activation='relu'))
ann.add(tf.keras.layers.Dense(units=100, activation='relu'))
ann.add(tf.keras.layers.Dense(units=100, activation='relu'))
ann.add(tf.keras.layers.Dense(units=100, activation='relu'))
ann.add(tf.keras.layers.Dense(units=100, activation='relu'))
ann.add(tf.keras.layers.Dense(units=100, activation='relu'))

ann.add(tf.keras.layers.Dense(units=1, activation='sigmoid'))

ann.compile(optimizer = 'adam', loss = 'binary_crossentropy', metrics = ['accuracy'])

ann.fit(X_train, y_train, batch_size = 32, epochs = 500)

Epoch 1/500
23/23 [==============================] - 0s 4ms/step - loss: 0.6262 - accuracy: 0.6475
Epoch 2/500
23/23 [==============================] - 0s 3ms/step - loss: 0.4847 - accuracy: 0.8020
Epoch 3/500
23/23 [==============================] - 0s 3ms/step - loss: 0.4254 - accuracy: 0.8272
Epoch 4/500
23/23 [==============================] - 0s 3ms/step - loss: 0.4251 - accuracy: 0.8244
Epoch 5/500
23/23 [==============================] - 0s 4ms/step - loss: 0.4163 - accuracy: 0.8357
Epoch 6/500
23/23 [==============================] - 0s 4ms/step - loss: 0.3946 - accuracy: 0.8413
Epoch 7/500
23/23 [==============================] - 0s 4ms/step - loss: 0.3856 - accuracy: 0.8469
Epoch 8/500
23/23 [==============================] - 0s 4ms/step - loss: 0.4062 - accuracy: 0.8329
Epoch 9/500
23/23 [==============================] - 0s 3ms/step - loss: 0.4154 - accuracy: 0.8329
Epoch 10/500
23/23 [==============================] - 0s 3ms/step - loss: 0.3944 - accuracy: 0.8371
Epoch 11/

Predicting on the split data:

In [ ]:
y_pred = ann.predict(X_test)
y_pred = (y_pred > 0.5)
y_pred = pd.DataFrame(y_pred)

In [ ]:
from sklearn.preprocessing import LabelEncoder
le = LabelEncoder()
y = le.fit_transform(y_pred.values) 

In [ ]:
accuracy = sum((y == y_test))/y.shape[0]
accuracy

0.8156424581005587

### A smaller model

In [ ]:
ann2 = tf.keras.models.Sequential()

ann2.add(tf.keras.layers.Dense(units=10, activation='relu', input_dim = 8))

ann2.add(tf.keras.layers.Dense(units=10, activation='relu'))

ann2.add(tf.keras.layers.Dense(units=1, activation='sigmoid'))

ann2.compile(optimizer = 'adam', loss = 'binary_crossentropy', metrics = ['accuracy'])

#ann2.fit(x, y, batch_size = 32, epochs = 500)

Predicting on the split data:

In [ ]:
y_pred2 = ann2.predict(X_test)
y_pred2 = (y_pred2 > 0.5)
y_pred2 = pd.DataFrame(y_pred2)

In [ ]:
y2 = le.fit_transform(y_pred2.values) 
accuracy2 = sum((y2 == y_test))/y2.shape[0]
accuracy2

0.8659217877094972

The much smaller model has an accuracy of 86.76% on the training data while the large model has an accuracy of 92.13%, seemingly a large difference. However when we use the two models to make predictions on a new set of data the smaller model is at 86.59% accuracy and the large is 81.56%. Clearly the smaller model extrapoltes better to new data.

### Saving to a .csv

In [ ]:
y = pd.DataFrame(y)
y.to_csv("ANNtest.csv", index = False)